// @ant-design/x/es/sources/components/CarouselCard.js
import { LeftOutlined, RightOutlined } from '@ant-design/icons';
import { Carousel } from 'antd';
import { clsx } from 'clsx';
import React, { useEffect } from 'react';
const CarouselCard = props => {
  const {
    prefixCls,
    items,
    activeKey,
    className,
    style
  } = props;
  const compCls = `${prefixCls}-carousel`;
  const [slide, setSlide] = React.useState(Math.max(0, items?.findIndex(({
    key
  }) => key === activeKey) ?? 0));
  const carouselRef = React.useRef(null);
  const handleClick = item => {
    item.url && window.open(item.url, '_blank', 'noopener,noreferrer');
    props.onClick?.(item);
  };
  useEffect(() => {
    if (carouselRef.current) {
      carouselRef.current.goTo(slide, false);
    }
  }, [slide]);
  return /*#__PURE__*/React.createElement("div", {
    style: style,
    className: clsx(`${compCls}-wrapper`, className)
  }, /*#__PURE__*/React.createElement("div", {
    className: `${compCls}-title`
  }, /*#__PURE__*/React.createElement("div", {
    className: `${compCls}-btn-wrapper`
  }, /*#__PURE__*/React.createElement("span", {
    className: clsx(`${compCls}-btn`, `${compCls}-left-btn`, {
      [`${compCls}-btn-disabled`]: slide === 0
    }),
    onClick: () => {
      if (slide > 0) {
        setSlide(pre => pre - 1);
      }
    }
  }, /*#__PURE__*/React.createElement(LeftOutlined, null)), /*#__PURE__*/React.createElement("span", {
    className: clsx(`${compCls}-btn`, `${compCls}-right-btn`, {
      [`${compCls}-btn-disabled`]: slide === (items?.length || 1) - 1
    }),
    onClick: () => {
      if (slide < (items?.length || 1) - 1) {
        setSlide(pre => pre + 1);
      }
    }
  }, /*#__PURE__*/React.createElement(RightOutlined, null))), /*#__PURE__*/React.createElement("div", {
    className: `${compCls}-page`
  }, `${slide + 1}/${items?.length || 1}`)), /*#__PURE__*/React.createElement(Carousel, {
    className: compCls,
    ref: carouselRef,
    arrows: false,
    infinite: false,
    dots: false,
    beforeChange: (_, nextSlide) => setSlide(nextSlide)
  }, items?.map((item, index) => /*#__PURE__*/React.createElement("div", {
    key: item.key || index,
    className: `${compCls}-item`,
    onClick: () => handleClick(item)
  }, /*#__PURE__*/React.createElement("div", {
    className: `${compCls}-item-title-wrapper`
  }, item.icon && /*#__PURE__*/React.createElement("span", {
    className: `${compCls}-item-icon`
  }, item.icon), /*#__PURE__*/React.createElement("span", {
    className: `${compCls}-item-title`
  }, item.title)), item.description && /*#__PURE__*/React.createElement("div", {
    className: `${compCls}-item-description`
  }, item.description)))));
};
export default CarouselCard;
